File: 0048_Rotate_Image/0048_Rotate_Image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        def reverse(matrix, N):
            N = len(matrix)
            for i in range(N):
                lo, hi = 0, N-1
                while lo < hi:
                    matrix[i][lo], matrix[i][hi] = matrix[i][hi], matrix[i][lo]
                    lo += 1
                    hi -= 1
            return matrix

        N = len(matrix)
        matrix = reverse(matrix, N)

        for i in range(0, N-1):
            for j in range(0, N-i-1):
                matrix[i][j], matrix[N-1-j][N-1-i] = matrix[N-1-j][N-1-i], matrix[i][j]
        return matrix

File: 0048_Rotate_Image/0048_Rotate_Image.cc
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int N = matrix.size();
        int lo = 0;
        int hi = N-1;
        int tmp = 0;
        for (int i = 0; i < N; ++i) {
            lo = 0;
            hi = N-1;
            while (lo < hi) {
                tmp = matrix[i][lo];
                matrix[i][lo] = matrix[i][hi];
                matrix[i][hi] = tmp;
                lo += 1;
                hi -= 1;
            }
        }

        for (int i = 0; i < N-1; ++i) {
            for (int j = 0; j < N-i-1; ++j) {
                tmp = matrix[N-1-j][N-1-i];
                matrix[N-1-j][N-1-i] = matrix[i][j];
                matrix[i][j] = tmp;
            }
        }
    }
};